File: calculate-business.lua
#!/bin/lua
--A script to calculate total business earnings (and losses) in D&D 5e
math.randomseed(os.time())

--just replace all of this using lua_cliargs :/
--also make functions local
function print_help(error_message)
	if error_message ~= nil then
		io.write("ERROR: ", error_message, "\n")
	end
	io.write("Usage: calculate-business.lua [arguments]\n Valid arguments are:\n\t-h\t\t\t| Print these options\n\t-c <weekly_cost>\t| Sets the weekly cost of upkeep for the business (required)\n\t-w <number_of_weeks>\t| Sets the amount of weeks to roll for; defaults to 1 and conflicts with -f\n\t-f <path_to_file>\t| Defines a file of rolls to use instead of generating them automatically; conflicts with -w\n")
	os.exit()
end

function match_whole_pattern(input, pattern)
	return string.find(input, "%f[%w-]" .. pattern .. "%f[^%w-]")
end

if #arg == 0 or string.match(table.unpack(arg), "-h") then print_help() end
local flags_paired = {"-b", "-c", "-d", "-f", "-w", "--divisor"}
local params = {}
for i = 1, #arg, 2 do
	params[arg[i]] = arg[i+1]
end

for k, v in pairs(params) do --ensure that all arguments have values and are correct
	local match_unrecognized = true
	local match_novalue = false
	for _, w in ipairs(flags_paired) do
		if match_whole_pattern(w, k) ~= nil then match_unrecognized = false end
		if match_whole_pattern(w, v) ~= nil then match_novalue = true end
	end
	if match_unrecognized == true then print_help("Argument not recognized") end
	if match_novalue == true then print_help("An argument is missing a value") end
end

if params["-b"] == nil then
	print_help("No balance provided")
elseif params["-c"] == nil then
	print_help("No weekly cost provided")
elseif params["-w"] ~= nil and params["-f"] ~= nil then
	print_help("-f and -w are incompatible with one another")
elseif params["-d"] ~= nil and params["-f"] ~= nil then
	print_help("-d and -f are incompatible with one another")
elseif params["-w"] ~= nil and params["-d"] ~= nil then
	print_help("-d and -w are incompatible with one another")
end

local balance = tonumber(params["-b"])
local weekly_cost = tonumber(params["-c"])
local weeks = params["-w"] or 1
local days = tonumber(params["-d"]) or (tonumber(weeks) * 7)
local file = params["-f"]
local divisor = tonumber(params["--divisor"]) or 1
--local verbose = params["-v"] or false

function file_exists(file)
	local f = io.open(file, "rb")
	if f then f:close() end
	return f ~= nil
end

function lines_from(file)
	if not file_exists(file) then return {} end
	local lines = {}
	for line in io.lines(file) do
		lines[#lines + 1] = line
	end
	return lines
end

function roll(sides, number)
	if number == nil then number = 1 end
	result = math.random(sides)
	return result
end

function roll_bonus()
	--track # of days up to 30
	local roll_count = 0
	return function ()
		if roll_count < 30 then roll_count = roll_count + 1 end
		return roll_count --closure to keep track of bonus up to 30
	end
end

function debt_penalty() --use an instance of this closure in check_rolls() to dynamically update a new total bonus
	local penalty_count = 0
	return function (reset)
		if reset == false then
			penalty_count = 0
			reset = nil
		else
			penalty_count = penalty_count + 10
			return penalty_count
		end
	end
end

function roll_days()
	local rolls = {}
	for i = 1, days do
		rolls[i] = roll(100) 
	end
	return rolls
end

function line_days()
	local lines = lines_from(file)
	for i, v in ipairs(lines) do
		lines[i] = tonumber(v)
	end
	return lines
end

function check_rolls()
	local sum = 0
	local cost_profit = 0
	local values = {}
	if file ~= nil then
		values = line_days()
	else
		values = roll_days()
	end
	--if verbose == true then io.write("Values: ", table.unpack(values), "\n") end
	local bonus = roll_bonus()
	local penalty = debt_penalty()
	for _, v in ipairs(values) do --table from dmg p.129
		value = v + bonus()
		if balance <= 0 then value = value - penalty() else penalty(false) end --reset penalty to roll if debt can be paid (balance goes back above 0)
		if value <= 0 then value = 1 end
		if value >= 1 and value <= 20 then
			cost_profit = (weekly_cost * -1.5)
			sum = sum + cost_profit
		elseif value >= 21 and value <= 30 then
			cost_profit = (weekly_cost * -1)
			sum = sum + cost_profit
		elseif value >= 31 and value <= 40 then
			cost_profit = (weekly_cost * -0.5)
			sum = sum + cost_profit 
		elseif value >= 41 and value <= 60 then
			goto continue
		elseif value >= 61 and value <= 80 then
			cost_profit = ((roll(6) * 5)/divisor)
			sum = sum + cost_profit
		elseif value >= 81 and value <= 90 then
			cost_profit = ((roll(8, 2) * 5)/divisor)
			sum = sum + cost_profit 
		elseif value >= 91 then
			cost_profit = ((roll(10, 3) * 5)/divisor)
			sum = sum + cost_profit 
		end
		balance = balance + cost_profit
		::continue::
		--if verbose == true then io.write("Current iteration: ", _, "\tCurrent value: ", value, "\tCurrent sum: ", sum, "\tCurrent balance: ", balance, "\tCurrent cost or profit: ", cost_profit, "\n") end
	end
	return sum, balance
end

local total, total_balance = check_rolls()

if total < 0 then
	io.write("This business has lost ", total, "gp. Your new balance is ", total_balance, "gp.\n")
	os.exit()
else
	io.write("This business has made ", total, "gp. Your new balance is ", total_balance, "gp.\n")
	os.exit()
end

File: calculate-business.py
#!/usr/bin/env python
import argparse
import re
import random

class Business:
    def __init__(self, balance, upkeep, time, divisor):
        self._balance = balance
        self._upkeep = upkeep
        
        time_match = re.search(r"^([0-9])+([dwmy]){1}$", time, flags=re.IGNORECASE)
        try:
            if not time_match.group(1):
                raise ValueError("No time value supplied")
            self._time = int(time_match.group(1))
        except (ValueError, AttributeError):
            print(parser.error("argument TIME: value is invalid"))
        try:
            if not time_match.group(2):
                raise ValueError("Interval not defined or incorrect")
            self._interval = time_match.group(2)
        except (ValueError, AttributeError):
            print(parser.error("argument TIME: value is invalid"))
        
        match self._interval:
            case "d":
                self._interval = 1
            case "w":
                self._interval = 7
            case "m":
                self._interval = 30
            case "y":
                self._interval = 365
            case _:
                raise ValueError("Interval incorrect; must be d, w, m, or y")
        self._time *= self._interval

        self._divisor = divisor
        #Initialize private variables to 0
        self._bonus = 0
        self._penalty = 0
        self._profit = 0
    
    def _roll(self, number=1, sides=6):
        i = 0
        total = 0
        while i < number:
            total = total + random.randint(1, sides)
            i += 1
        total = total - self._penalty + self._bonus
        if total < 1:
            total = 1
        return total

    def _bonus_increment(self, reset=False):
        if reset == True:
            self._bonus = 0
        if self._bonus < 30:
            self._bonus += 1

    def _penalty_check(self, reset=False):
        if reset == True:
            self._penalty = 0
        if self._balance < 0:
            self._penalty += 10
    
    def run(self):
        i = 0
        while i < self._time:
            roll = self._roll(sides=100)
            profit_old = self._profit
            if 1 <= roll <= 20:
                iteration_profit = self._upkeep * -1.5
            elif 21 <= roll <= 30:
                iteration_profit = self._upkeep * -1
            elif 31 <= roll <= 40:
                iteration_profit = self._upkeep * -0.5
            elif 41 <= roll <= 60:
                iteration_profit = 0
            elif 61 <= roll <= 80:
                iteration_profit = (self._roll() * 5 )/self._divisor
            elif 81 <= roll <= 90:
                iteration_profit = (self._roll(number=2, sides=8) * 5)/self._divisor
            elif 91 <= roll:
                iteration_profit = (self._roll(number=3, sides=10) * 5)/self._divisor
            
            self._profit = round(self._profit + iteration_profit, 2)
            self._balance = round(self._balance + iteration_profit, 2) # Balance is updated each iteration to check for penalty
            self._bonus_increment()
            if self._profit < profit_old:
                self._penalty_check()
            else:
                # The DMG table is not clear if the cumulative -10 penalty is kept even if unpaid debts are paid off.
                # In my opinion, since the penalty is cumulative (which is also subject to debate), it would be unfair not to reset the penalty
                # once the character's balance is above 0.
                self._penalty_check(reset=True)
            
            i += 1
        # Reset bonus and penalty to 0 after run() has completed for future-proofing
        # (e.g., if this class is imported to another program, it could support being run multiple times with persistent profits and balance)
        self._bonus_increment(reset=True)
        self._penalty_check(reset=True)

    @property
    def balance(self):
        return self._balance

    @property
    def upkeep(self):
        return self._upkeep

    @property
    def time(self):
        return self._time

    @property
    def interval(self):
        return self._interval

    @property
    def divisor(self):
        return self._divisor

    @property
    def profit(self):
        return self._profit

parser = argparse.ArgumentParser(description="A script to calculate total business earnings (and losses) in D&D 5e")
parser.add_argument("balance", type=float, help="Character's current amount of gold (can include silver and copper as decimals)", metavar="BALANCE")
parser.add_argument("upkeep", type=float, help="Daily cost of business in gold (can include silver and copper as decimals)", metavar="UPKEEP")
parser.add_argument("time", type=str, help="Amount of time spent running business. Format: number[d,w,m,y], where d is days, w is weeks, m is months, and y is years.", metavar="TIME")
parser.add_argument("--divisor", type=int, help="If your campaign balances gold prices differently, you can specify how much to divide profits by", default=1)
args = parser.parse_args()

def main():
    business = Business(args.balance, args.upkeep, args.time, args.divisor)
    old_balance = business.balance
    business.run()
    if business.balance < old_balance:
        print(f"This business has lost {business.profit}gp. The new balance is {business.balance}gp.")
    elif business.balance == old_balance:
        print(f"This business broke even. The balance is still {business.balance}.")
    else:
        print(f"This business has earned {business.profit}gp. The new balance is {business.balance}gp.")

if __name__ == "__main__":
    main()
